Sources/Transformer.swift

```diff
@@ -0,0 +1,177 @@
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+import AlgoliaSearch
+
+/// A Transformable data source
+public protocol Transformable {
+    associatedtype IResults
+    associatedtype IParameters
+    
+    typealias SearchResultsHandler = (_ results: IResults?, _ error: Error?) -> Void
+    
+    // Search operation
+    func search(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia params to custom backend params.
+    func map(query: Query) -> IParameters
+    
+    // Transforms the custom backend result to an Algolia result.
+    func map(results: IResults) -> SearchResults
+    
+    
+    // Disjunctive Search operation
+    func searchDisjunctiveFaceting(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia disjunctive params to custom backend params.
+    func map(query: Query, disjunctiveFacets: [String], refinements: [String : [String]]) -> IParameters
+    
+    
+    // Search for facet value operation
+    func searchForFacetValues(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia facet value params to custom backend params.
+    func map(query: Query?, facetName: String, matching text: String) -> IParameters
+}
+
+/// Base class used to implement a custom backend with Algolia
+/// You need to make sure to implement the following method:
+/// - search(query:searchResultsHandler:)
+/// - map(query:)
+/// - map(results:)
+open class SearchTransformer<Parameters, Results>: Transformable, Searchable {
+    
+    public typealias IParameters = Parameters
+    public typealias IResults = Results
+    
+    public init() {}
+    
+    // Transformable protocol
+    
+    open func search(_ query: Parameters, searchResultsHandler: @escaping SearchResultsHandler) {
+        fatalError("make sure to override search(query:searchResultsHandler:) for custom backend")
+    }
+    
+    open func map(query: Query) -> Parameters {
+        fatalError("make sure to override map(query:) for custom backend")
+    }
+    
+    open func map(results: Results) -> SearchResults {
+        fatalError("make sure to override map(results:) for custom backend")
```

Comment:
Same comment as `map(query: Query) -> Parameters`.

Reply:
check answer above

Reply:
I didn't notice but shouldn't it be `Result` instead of `Results`? (singular)

The function maps a result at a time (like `map(query: Query)`).

Reply:
mhmm that's interesting. If we want to change that, then we also should change from `SearchResults` to `SearchResult`.

I am personally fine with both.. I think `Results` works too and is not misleading. 

WDYT @PLNech @robertmogos ? we can deprecate `SearchResults` if we prefer `SearchResult`(same for `Results`)

Reply:
I would weight in for `SearchResults`: you usually hear `the search results` (as the objects returned for this query) more often that `search result` (the result of searching, which is a set of objects)

Reply:
The web team actually went for "response" instead of "result".

But this naming in Swift is internal, right?